sheet08/sheet08solutions.ipynb

```diff
@@ -327,373 +264,223 @@
+    "        for layer in self.layers:\n",
+    "            x = layer.activate(x)\n",
+    "        return x\n",
     "\n",
-    "    def adaption(self, X, t):\n",
-    "        \"\"\"\n",
-    "        Trains the perceptron. Adjusts the weights according to \n",
-    "        the learning rate and the error between the activation and the \n",
-    "        label (delta).\n",
+    "    def adapt(self, x, t, rate=0.03):\n",
+    "        \"\"\"Adapt the whole network given an input and expected output.\n",
     "\n",
     "        Args:\n",
-    "            X   the data point, should be a numpy\n",
-    "                array or a 1xN numpy matrix\n",
-    "            t   the label\n",
+    "            x: Vector of input values.\n",
+    "            t: Vector of target values (expected outputs).\n",
+    "            rate: Learning rate.\n",
     "        \"\"\"\n",
-    "        self.w += self.epsilon * (t - self.activation(X)) * np.append(1, X)\n",
+    "        # Activate each layer and collect intermediate outputs.\n",
+    "        # outputs = []  # TODO\n",
+    "        outputs = [x]\n",
+    "        for layer in self.layers:\n",
+    "            outputs.append(layer.activate(outputs[-1]))\n",
+    "\n",
+    "        # Calculate error between t and network output.\n",
+    "        # e = np.array([])  # TODO\n",
+    "        e = t - outputs[-1]\n",
     "\n",
-    "    def __repr__(self):\n",
-    "        return '{}({}, {})'.format(self.__class__.__name__, len(self.w) - 1, self.epsilon)"
+    "        # Backpropagate error through the network computing\n",
+    "        # intermediate delta and adapting each layer.\n",
+    "        for k in range(len(self.layers) - 1, 0, -1):\n",
+    "            delta = (outputs[k + 1] * (1 - outputs[k + 1])) * e\n",
+    "            weights = self.layers[k].get_weights_matrix()\n",
+    "            e = (weights @ delta)[1:]\n",
+    "            self.layers[k].adapt(outputs[k], delta, rate)"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "This cell instantiates a BasePerceptron (`input_dim` is taken from above, where the data is created) and trains and tests it. The multilayer perceptron will behave exactly the same."
+    "### Classification"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Problem Definition\n",
+    "Before we start, we need a problem to solve. In the following cell we first generate some three dimensional data (= $\\text{input_dim}$) between 0 and 1 and label all data according to a binary classification: If the data is close to the center (radius < 2.5), it belongs to one class, if it is further away from the center it belongs to the other class.\n",
+    "\n",
+    "In the cell below we visualize the data set."
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
    "metadata": {
-    "collapsed": false
+    "collapsed": true
```

Comment:
I just blank out all metadata where possible, but I think `collapsed` for a visualization cell is not always a good idea, even though it's just one click.